### PR TITLE
Promotion quiz now updates the user's game stage

### PR DESCRIPTION
**Pull request by lzoebaker**
If the user passes the promotion quiz, their game stage is updated. If the user fails the promotion quiz, their game stage stays the same. 
If the user is on the final stage of a path and they somehow end up on the promotion quiz page, a message is printed to the screen instead of the questions and choices. 

** After talking with Will, filed a bug issue to change the generic Exceptions that the user auth/ player database throws to a more specific custom LoggedOutException. This will be fixed in a later PR.

### CODE REVIEW

File: src/main/java/com/google/sps/data/GameStage.java [hunk 1]
```diff
@@ -18,6 +18,8 @@
    * Game Stages are levels on the gameboard. Each level can be accessed by a player, but only one
    * at a time.
    */
+  private static final String DEFAULT_NEXTID_IF_FINAL_STAGE = "nonextid";
```
Comment (lattai) [inline src/main/java/com/google/sps/data/GameStage.java:21]: Its a very funky string, if you're able to hard code it and change it, maybe camel case or with dashes?

File: src/main/java/com/google/sps/data/GameStage.java [hunk 2]
```diff
@@ -18,6 +18,8 @@
    * Game Stages are levels on the gameboard. Each level can be accessed by a player, but only one
    * at a time.
    */
+  private static final String DEFAULT_NEXTID_IF_FINAL_STAGE = "noNextId";
+
```
Comment (menynexi) [inline src/main/java/com/google/sps/data/GameStage.java:22]: Weird line

Comment (lzoebaker) [reply src/main/java/com/google/sps/data/GameStage.java:22]: I think the linter wants me to have a line between my static final variables and my other variables.

File: src/main/java/com/google/sps/data/GameStage.java [hunk 3]
```diff
@@ -54,7 +56,7 @@ public GameStage(
     this.id = id;
     this.quizKey = quizKey;
     this.isFinalStage = isFinalStage;
-    this.nextStageId = nextStageId;
+    this.nextStageId = isFinalStage ? DEFAULT_NEXTID_IF_FINAL_STAGE : nextStageId;
```
Comment (WillHipschman) [inline src/main/java/com/google/sps/data/GameStage.java:59]: NEXTID-->NEXT_ID

File: src/main/java/com/google/sps/servlets/PromotionQuizServlet.java [hunk 4]
```diff
@@ -21,36 +26,44 @@
...
-  private static final String HARDCODED_QUIZ_KEY = "softwareEngineerLevel1";
   private static final String PROMOTED_MESSAGE =
       "Congratulations, you passed the quiz and were promoted!";
   private static final String NOT_PROMOTED_MESSAGE =
       "You did not pass the quiz. Study the content and try again later";
-  private static final Double CORRECT_NEEDED_THRESHOLD = 0.5;
-
-  private static final Gson gson = new Gson();
+  private static final String IS_FINAL_STAGE_MESSAGE =
+      "Congratulations! You reached the final stage! You may no longer be promoted in this path.";
```
Comment (WillHipschman) [inline src/main/java/com/google/sps/servlets/PromotionQuizServlet.java:34]: This is a bit of a weird message. What is the user supposed to do? Start over in another path? Stop playing? We should provide messaging that makes it clear what they should do next,

File: src/main/java/com/google/sps/servlets/PromotionQuizServlet.java [hunk 5]
```diff
@@ -60,18 +73,42 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
-    if (isPromoted) {
-      // TODO: https://github.com/googleinterns/step144-2020/issues/88
-      // then get the next id by calling currentGameStage.getNextStageId()
-      // then set the players gameStage to the next id
+    try {
+      this.quizQuestions = getQuizQuestions(response);
+      Boolean isPromoted = handleQuizSubmission(request);
+      PromotionMessage promotionMessage =
+          new PromotionMessage(isPromoted, isPromoted ? PROMOTED_MESSAGE : NOT_PROMOTED_MESSAGE);
+      if (isPromoted) {
```
Comment (WillHipschman) [inline src/main/java/com/google/sps/servlets/PromotionQuizServlet.java:81]: What does the user do if they're not promoted? Do they redo the quiz? Do they go through the content again?

File: src/main/java/com/google/sps/servlets/PromotionQuizServlet.java [hunk 6]
```diff
@@ -60,18 +73,42 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    String currentGameStageId = this.playerDatabase.getEntityCurrentPageID();
+    return this.gameStageDatabase.getGameStage(currentGameStageId);
+  }
+
+  private boolean isUserOnFinalStage() throws Exception {
+    GameStage currentGameStage = getCurrentGameStage();
+    return currentGameStage.isFinalStage();
+  }
+
+  private ArrayList<QuizQuestion> getQuizQuestions(HttpServletResponse response)
```
Comment (WillHipschman) [inline src/main/java/com/google/sps/servlets/PromotionQuizServlet.java:102]: Can this just be List<QuizQuestion>?

File: src/main/java/com/google/sps/servlets/PromotionQuizServlet.java [hunk 7]
```diff
@@ -21,36 +26,44 @@
...
-    response.getWriter().println(quizQuestionsJson);
+    try {
+      if (isUserOnFinalStage()) {
+        response.getWriter().println(gson.toJson(IS_FINAL_STAGE_MESSAGE));
+      } else {
+        this.quizQuestions = getQuizQuestions(response);
+        String quizQuestionsJson = gson.toJson(quizQuestions);
+        response.getWriter().println(quizQuestionsJson);
+      }
+    } catch (Exception e) {
```
Comment (WillHipschman) [inline src/main/java/com/google/sps/servlets/PromotionQuizServlet.java:64]: Here (and everywhere you catch Exception) we should catch a more specific question. i.e. If we get a NullPointerException (a bug in the code) we shouldn't treat that as a user not being logged in, we should just not handle it.

File: src/test/java/com/google/sps/PromotionQuizTest.java [hunk 8]
```diff
@@ -33,25 +43,65 @@
 /** Tests the PromotionQuiz servlet and its interactions with QuestionDatabase */
 @RunWith(JUnit4.class)
 public final class PromotionQuizTest {
+  // creating a current user for testing
+  private static final String NAME = "Linda";
+  private static final String EMAIL = "Linda@email.com";
+  private static final String AUTH_DOMAIN = "email.com";
+  private static final String CURR_USER_ID = "testidme";
```
Comment (WillHipschman) [inline src/test/java/com/google/sps/PromotionQuizTest.java:50]: CURRENT_USER_ID is easier to read and not much longer ;)

File: src/main/java/com/google/sps/data/GameStage.java (hunk 2 above)
Comment (DirtGrubDylan) [reply src/main/java/com/google/sps/data/GameStage.java:22]: it does

File: src/main/java/com/google/sps/servlets/PromotionQuizServlet.java (hunk 4 above)
Comment (lzoebaker) [reply src/main/java/com/google/sps/servlets/PromotionQuizServlet.java:34]: I think this message should almost never be displayed in normal game flow, only if the user manages to get to promotionquiz.html on their own. However, you make a good point about providing a next step. If a future PR, I will be handling how all these messages will be displayed in the front end.

File: src/main/java/com/google/sps/servlets/PromotionQuizServlet.java (hunk 7 above)
Comment (lzoebaker) [reply src/main/java/com/google/sps/servlets/PromotionQuizServlet.java:64]: So the only problem is that Leah's code throws an Exception (generic), whenever a user is logged out. So I am unsure as how to catch it more specifically.

File: src/main/java/com/google/sps/servlets/PromotionQuizServlet.java (hunk 5 above)
Comment (lzoebaker) [reply src/main/java/com/google/sps/servlets/PromotionQuizServlet.java:81]: Their game stage will be unchanged, and they will be redirected back to the game stage page to go through the content again. In a future PR, I will be linking the promotion quiz functionality to the game stage template.

**Pull request merged**